include/wil/result_macros.h

```diff
@@ -1340,10 +1339,14 @@ namespace wil
                                bool fWantDebugString, _Out_writes_(debugStringSizeChars) _Post_z_ PWSTR debugString, _Pre_satisfies_(debugStringSizeChars > 0) size_t debugStringSizeChars,
                                _Out_writes_(callContextStringSizeChars) _Post_z_ PSTR callContextString, _Pre_satisfies_(callContextStringSizeChars > 0) size_t callContextStringSizeChars,
                                _Out_ FailureInfo *failure) WI_NOEXCEPT;
-        _When_((type == FailureType::Exception) || (type == FailureType::FailFast), _Analysis_noreturn_) __declspec(noinline) inline void ReportFailure(__R_FN_PARAMS_FULL, FailureType type, HRESULT hr, _In_opt_ PCWSTR message = nullptr, ReportFailureOptions options = ReportFailureOptions::None);
-        _When_((type == FailureType::Exception) || (type == FailureType::FailFast), _Analysis_noreturn_) inline void ReportFailure_ReplaceMsg(__R_FN_PARAMS_FULL, FailureType type, HRESULT hr, _Printf_format_string_ PCSTR formatString, ...);
-        _When_((type == FailureType::Exception) || (type == FailureType::FailFast), _Analysis_noreturn_) __declspec(noinline) inline void ReportFailure_Hr(__R_FN_PARAMS_FULL, FailureType type, HRESULT hr);
-        _When_((type == FailureType::Exception) || (type == FailureType::FailFast), _Analysis_noreturn_) __declspec(noinline) inline HRESULT ReportFailure_CaughtException(__R_FN_PARAMS_FULL, FailureType type, SupportedExceptions supported = SupportedExceptions::Default);
+        template<FailureType, bool = false>                       

```

Comment:
>FailureType [](start = 17, length = 11)

The `FailureType` template parameter needs a name for the SAL below (assuming SAL can even handle template arguments)

Reply:
Is there a way to validate SAL? Even without 'type' being defined this compiled without issue. I can't find any resources that mention SAL and template parameters. I'd imagine they probably wouldn't play nice together so lacking a way to validate the SAL I'll just remove it.

Reply:
The compiler doesn't really care about SAL (one of the many reasons I prefer not to use it). @jonwis may know more since he recently did some work to fix some of them up. I believe you ran run code analysis in VS or something like that

Reply:
I suspect SAL does not do template params.
can we move these to the specializations of the functions and not make them conditional?